sklearn/preprocessing/imputation.py

```diff
@@ -366,13 +397,37 @@ def transform(self, X):
 
             mask = _get_mask(X, self.missing_values)
             n_missing = np.sum(mask, axis=self.axis)
-            values = np.repeat(valid_statistics, n_missing)
 
-            if self.axis == 0:
-                coordinates = np.where(mask.transpose())[::-1]
+            if self.strategy == 'knn':
+                if self.axis == 1:
+                    X = X.transpose()
+                    mask = mask.transpose()
+                    statistics = statistics.transpose()
+                missing_index = np.where(mask.any(1))[0]
+                for i, row in zip(missing_index, X[missing_index]):
+                    col_index = np.where(np.logical_not(np.isnan(row)))[0]
+                    impute_index = np.where(np.isnan(row))[0]
+                    neigh = NearestNeighbors(self.kneighbor)
```

Comment:
I think having euclidean distances only would be fine. But I'm not sure how to do the interface for your suggestion. For large datasets we might want to give an option to the user to do something else.